### ISSUE TITLE
Latest plugin release causes segfault

### ISSUE THREAD

**Issue by rrthomas** [jgeboski/purple-facebook#200]
Having upgraded to the latest release, pidgin crashes on startup:

    (gdb) where
    #0  __GI___libc_free (mem=0x20) at malloc.c:2929
    #1  0x00007f1d0d7c81e3 in ?? () from /usr/lib/purple-2/libfacebook.so
    #2  0x00007f1d0d7c6e8d in ?? () from /usr/lib/purple-2/libfacebook.so
    #3  0x00007f1d0d7d0e2b in ?? () from /usr/lib/purple-2/libfacebook.so
    #4  0x00007f1d0d7d2dbf in ?? () from /usr/lib/purple-2/libfacebook.so
    #5  0x00007f1d0d7d35e0 in ?? () from /usr/lib/purple-2/libfacebook.so
    #6  0x000055bea59d74ce in ?? ()
    #7  0x00007f1d1f200ce5 in g_main_dispatch (context=0x55bea7b50bc0) at /build/buildd/glib2.0-2.40.2/./glib/gmain.c:3064
    #8  g_main_context_dispatch (context=context@entry=0x55bea7b50bc0) at /build/buildd/glib2.0-2.40.2/./glib/gmain.c:3663
    #9  0x00007f1d1f201048 in g_main_context_iterate (context=0x55bea7b50bc0, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimised out>)
        at /build/buildd/glib2.0-2.40.2/./glib/gmain.c:3734
    #10 0x00007f1d1f20130a in g_main_loop_run (loop=0x55bea8ce84e0) at /build/buildd/glib2.0-2.40.2/./glib/gmain.c:3928
    #11 0x00007f1d2048a447 in gtk_main () from /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0
    #12 0x000055bea599e369 in main ()

with the deb for 20160118~909c36b~25a255f32eee~100

**Comment by rrthomas** [comment]
Tried downloading to release 99, same problem.

**Comment by rrthomas** [comment]
Tried re-upgrading to release 100, then manually removing account from accounts.xml and recreating it. Pidgin starts up, connects to Facebook, shows a group chat, then crashes.

If I disable the Facebook account before it connects, Pidgin doesn't crash any more.

**Comment by rrthomas** [comment]
I wondered whether this might be the same as #165, only before I recreated the account, I had disabled new group messages from showing.

**Comment by jgeboski** [comment]
Could you build the plugin with debugging information and install it locally? I have outlined the procedure below. This will require you uninstall the purple-facebook package (`sudo apt-get remove purple-facebook`). Sorry for the run around on this, I really need to setup `-dbg` packages, I just lack the time and motivation.

```
$ wget https://github.com/jgeboski/purple-facebook/releases/download/25a255f32eee/purple-facebook-25a255f32eee.tar.gz
$ tar xf purple-facebook-25a255f32eee.tar.gz
$ cd purple-facebook-25a255f32eee
$ CFLAGS="-g -O0" ./autogen.sh --enable-warnings
$ make
$ mkdir -p ~/.purple/plugins
$ cp pidgin/libpurple/protocols/facebook/.libs/libfacebook.so ~/.purple/plugins
```

Then obtain a [stack trace](See: https://github.com/jgeboski/purple-facebook/wiki/Debugging-Logs#obtaining-a-gdb-backtrace). Once you have obtained a stack trace, then `rm ~/.purple/plugins/libfacebook.so`, and reinstall bitlbee-facebook via `apt-get`.

**Comment by rrthomas** [comment]
Not to worry, the procedure is quick and clear (though I presume you mean "reinstall purple-facebook"?).

Stack trace:

    #0  __GI___libc_free (mem=0x32) at malloc.c:2929
    #1  0x00007f24f68f8777 in fb_cb_image (img=0x5638a5f53280, error=0x0)
        at facebook.c:448
    #2  0x00007f24f68f7784 in fb_data_image_cb (con=0x5638a6089220, 
        res=0x5638a62112d0, data=0x5638a5f53280) at data.c:547
    #3  0x00007f24f6905ed2 in purple_http_connection_terminate (hc=0x5638a6089220)
        at ../../http.c:1688
    #4  0x00007f24f6904cee in _purple_http_recv_loopbody (hc=0x5638a6089220, fd=34)
        at ../../http.c:1260
    #5  0x00007f24f6904d47 in _purple_http_recv (_hc=0x5638a6089220, fd=34, 
        cond=PURPLE_INPUT_READ) at ../../http.c:1271
    #6  0x00005638a2cd54ce in ?? ()
    #7  0x00007f24fddaece5 in g_main_dispatch (context=0x5638a498dbc0)
        at /build/buildd/glib2.0-2.40.2/./glib/gmain.c:3064
    #8  g_main_context_dispatch (context=context@entry=0x5638a498dbc0)
        at /build/buildd/glib2.0-2.40.2/./glib/gmain.c:3663
    #9  0x00007f24fddaf048 in g_main_context_iterate (context=0x5638a498dbc0, 
        block=block@entry=1, dispatch=dispatch@entry=1, self=<optimised out>)
        at /build/buildd/glib2.0-2.40.2/./glib/gmain.c:3734
    #10 0x00007f24fddaf30a in g_main_loop_run (loop=0x5638a59ed300)
        at /build/buildd/glib2.0-2.40.2/./glib/gmain.c:3928
    #11 0x00007f24ff038447 in gtk_main ()
       from /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0

**Comment by dequis** [comment]
I guess that makes this a duplicate of #165 

I'm not sure if we ever got a valgrind log from that one, so here: https://github.com/jgeboski/purple-facebook/wiki/Debugging-Logs#obtaining-a-valgrind-report

Most of that log file is private stuff but if you can just post the part where it says "==12345== Invalid read" or similar, followed by a couple of stack traces, that would be neat.

**Comment by rrthomas** [comment]
Here you go:

GDB backtrace:

    *** Error in `/usr/bin/pidgin': double free or corruption (!prev): 0x0000555556f65240 ***
    
    Program received signal SIGABRT, Aborted.
    0x00007ffff4779cc9 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
    56	../nptl/sysdeps/unix/sysv/linux/raise.c: No such file or directory.
    #0  0x00007ffff4779cc9 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
    #1  0x00007ffff477d0d8 in __GI_abort () at abort.c:89
    #2  0x00007ffff47b6394 in __libc_message (do_abort=do_abort@entry=1, fmt=fmt@entry=0x7ffff48c4b28 "*** Error in `%s': %s: 0x%s ***\n") at ../sysdeps/posix/libc_fatal.c:175
    #3  0x00007ffff47c266e in malloc_printerr (ptr=<optimised out>, str=0x7ffff48c4c10 "double free or corruption (!prev)", action=1) at malloc.c:4996
    #4  _int_free (av=<optimised out>, p=<optimised out>, have_lock=0) at malloc.c:3840
    #5  0x00007fffe0f06c43 in ?? () from /usr/lib/x86_64-linux-gnu/nss/libsoftokn3.so
    #6  0x00007fffe0ef4538 in ?? () from /usr/lib/x86_64-linux-gnu/nss/libsoftokn3.so
    #7  0x00007fffe618f35c in ?? () from /usr/lib/x86_64-linux-gnu/libnss3.so
    #8  0x00007fffe617d60b in PK11_DestroyContext () from /usr/lib/x86_64-linux-gnu/libnss3.so
    #9  0x00007fffe5cf0bf9 in ?? () from /usr/lib/x86_64-linux-gnu/libssl3.so
    #10 0x00007fffe5cfec6b in ?? () from /usr/lib/x86_64-linux-gnu/libssl3.so
    #11 0x00007fffe5d000b2 in ?? () from /usr/lib/x86_64-linux-gnu/libssl3.so
    #12 0x00007fffe5cf5bc0 in ?? () from /usr/lib/x86_64-linux-gnu/libssl3.so
    #13 0x00007fffe6488fd0 in ?? () from /usr/lib/purple-2/ssl-nss.so
    #14 0x00007ffff50c39a1 in purple_ssl_close () from /usr/lib/libpurple.so.0
    #15 0x00007fffedf040a2 in purple_socket_cancel (ps=0x555556c6ac00) at ../../purple-socket.c:371
    #16 0x00007fffedf04135 in purple_socket_destroy (ps=0x555556c6ac00) at ../../purple-socket.c:393
    #17 0x00007fffedefcbfe in purple_http_socket_close_free (hs=0x555556a75180) at ../../http.c:496
    #18 0x00007ffff53c56b8 in g_slist_foreach (list=<optimised out>, list@entry=0x555556d7c330, func=0x7fffedefcbb2 <purple_http_socket_close_free>, user_data=user_data@entry=0x0) at /build/buildd/glib2.0-2.40.2/./glib/gslist.c:877
    #19 0x00007ffff53c56db in g_slist_free_full (list=0x555556d7c330, free_func=<optimised out>) at /build/buildd/glib2.0-2.40.2/./glib/gslist.c:172
    #20 0x00007fffedf00b94 in purple_http_keepalive_host_free (_host=0x555556d46ec0) at ../../http.c:2079
    #21 0x00007ffff5397c67 in g_hash_table_remove_all_nodes (hash_table=0x555556d20580, notify=<optimised out>) at /build/buildd/glib2.0-2.40.2/./glib/ghash.c:503
    #22 0x00007ffff53989a1 in g_hash_table_remove_all (hash_table=hash_table@entry=0x555556d20580) at /build/buildd/glib2.0-2.40.2/./glib/ghash.c:1371
    #23 0x00007ffff5398a0e in g_hash_table_destroy (hash_table=0x555556d20580) at /build/buildd/glib2.0-2.40.2/./glib/ghash.c:1067
    #24 0x00007fffedf00c77 in purple_http_keepalive_pool_free (pool=0x555556b1a9e0) at ../../http.c:2111
    #25 0x00007fffedf00d37 in purple_http_keepalive_pool_unref (pool=0x555556b1a9e0) at ../../http.c:2135
    #26 0x00007fffedf016e7 in purple_http_request_free (request=0x555556ffbf00) at ../../http.c:2447
    #27 0x00007fffedf017d1 in purple_http_request_unref (request=0x555556ffbf00) at ../../http.c:2472
    #28 0x00007fffedeffcc5 in purple_http_connection_free (hc=0x555556cc2c00) at ../../http.c:1648
    #29 0x00007fffedeffede in purple_http_connection_terminate (hc=0x555556cc2c00) at ../../http.c:1690
    #30 0x00007fffedefecee in _purple_http_recv_loopbody (hc=0x555556cc2c00, fd=27) at ../../http.c:1260
    #31 0x00007fffedefed47 in _purple_http_recv (_hc=0x555556cc2c00, fd=27, cond=PURPLE_INPUT_READ) at ../../http.c:1271
    #32 0x00005555555c84ce in ?? ()
    #33 0x00007ffff53a8ce5 in g_main_dispatch (context=0x555555864bc0) at /build/buildd/glib2.0-2.40.2/./glib/gmain.c:3064
    #34 g_main_context_dispatch (context=context@entry=0x555555864bc0) at /build/buildd/glib2.0-2.40.2/./glib/gmain.c:3663
    #35 0x00007ffff53a9048 in g_main_context_iterate (context=0x555555864bc0, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimised out>) at /build/buildd/glib2.0-2.40.2/./glib/gmain.c:3734
    #36 0x00007ffff53a930a in g_main_loop_run (loop=0x5555568a9cd0) at /build/buildd/glib2.0-2.40.2/./glib/gmain.c:3928
    #37 0x00007ffff6632447 in gtk_main () from /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0
    #38 0x000055555558f369 in main ()

Valgrind errors:

    ==12004== Invalid read of size 8
    ==12004==    at 0xEF1792C: fb_api_message_free (api.c:3004)
    ==12004==    by 0x760E6B7: g_slist_foreach (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==12004==    by 0x760E6DA: g_slist_free_full (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==12004==    by 0xEF1A7EC: fb_cb_image (facebook.c:456)
    ==12004==    by 0xEF19783: fb_data_image_cb (data.c:547)
    ==12004==    by 0xEF27ED1: purple_http_connection_terminate (http.c:1688)
    ==12004==    by 0xEF26CED: _purple_http_recv_loopbody (http.c:1260)
    ==12004==    by 0xEF26D46: _purple_http_recv (http.c:1271)
    ==12004==    by 0x17C4CD: ??? (in /usr/bin/pidgin)
    ==12004==    by 0x75F1CE4: g_main_context_dispatch (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==12004==    by 0x75F2047: g_main_context_iterate.isra.24 (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==12004==    by 0x75F2309: g_main_loop_run (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==12004==    by 0x61F8446: gtk_main (in /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0.2400.23)
    ==12004==    by 0x143368: main (in /usr/bin/pidgin)
    ==12004==  Address 0x2cba4a50 is 16 bytes after a block of size 96 in arena "client"
    ==12004== 
    ==12004== Invalid free() / delete / delete[] / realloc()
    ==12004==    at 0x4C2BDEC: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
    ==12004==    by 0xEF17943: fb_api_message_free (api.c:3005)
    ==12004==    by 0x760E6B7: g_slist_foreach (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==12004==    by 0x760E6DA: g_slist_free_full (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==12004==    by 0xEF1A7EC: fb_cb_image (facebook.c:456)
    ==12004==    by 0xEF19783: fb_data_image_cb (data.c:547)
    ==12004==    by 0xEF27ED1: purple_http_connection_terminate (http.c:1688)
    ==12004==    by 0xEF26CED: _purple_http_recv_loopbody (http.c:1260)
    ==12004==    by 0xEF26D46: _purple_http_recv (http.c:1271)
    ==12004==    by 0x17C4CD: ??? (in /usr/bin/pidgin)
    ==12004==    by 0x75F1CE4: g_main_context_dispatch (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==12004==    by 0x75F2047: g_main_context_iterate.isra.24 (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==12004==    by 0x75F2309: g_main_loop_run (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==12004==    by 0x61F8446: gtk_main (in /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0.2400.23)
    ==12004==    by 0x143368: main (in /usr/bin/pidgin)
    ==12004==  Address 0x2cba4a30 is 80 bytes inside a block of size 88 free'd
    ==12004==    at 0x4C2BDEC: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
    ==12004==    by 0x795AA4B: xmlnode_free (in /usr/lib/libpurple.so.0.10.9)
    ==12004==    by 0x795AA4B: xmlnode_free (in /usr/lib/libpurple.so.0.10.9)
    ==12004==    by 0x795AA4B: xmlnode_free (in /usr/lib/libpurple.so.0.10.9)
    ==12004==    by 0x795AA4B: xmlnode_free (in /usr/lib/libpurple.so.0.10.9)
    ==12004==    by 0x795AA4B: xmlnode_free (in /usr/lib/libpurple.so.0.10.9)
    ==12004==    by 0x795AA4B: xmlnode_free (in /usr/lib/libpurple.so.0.10.9)
    ==12004==    by 0x78FCFC0: ??? (in /usr/lib/libpurple.so.0.10.9)
    ==12004==    by 0x78FD218: ??? (in /usr/lib/libpurple.so.0.10.9)
    ==12004==    by 0x75F2702: g_timeout_dispatch (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==12004==    by 0x75F1CE4: g_main_context_dispatch (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==12004==    by 0x75F2047: g_main_context_iterate.isra.24 (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==12004==    by 0x75F2309: g_main_loop_run (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==12004==    by 0x61F8446: gtk_main (in /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0.2400.23)
    ==12004==    by 0x143368: main (in /usr/bin/pidgin)
    ==12004==

**Comment by jgeboski** [comment]
How is this crashing, just upon connecting?

**Comment by rrthomas** [comment]
It seems to connect fine and then crash just after opening and displaying a group chat. The backtrace seems consistent with the theory that it crashes when trying to display an image. About 20 chat messages are visible in my case, from multiple contacts, but no images.

**Comment by dequis** [comment]
Can you install the pidgin-dbg package and repeat the last valgrind log?

**Comment by dequis** [comment]
```
==12004==    by 0x795AA4B: xmlnode_free (in /usr/lib/libpurple.so.0.10.9)
==12004==    by 0x78FCFC0: ??? (in /usr/lib/libpurple.so.0.10.9)
==12004==    by 0x78FD218: ??? (in /usr/lib/libpurple.so.0.10.9)
==12004==    by 0x75F2702: g_timeout_dispatch (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
```

Those two are probably `save_cb` and `purple_blist_sync` of blist.c, judging by the offsets from `xmlnode_free` in the `libpurple.so.0.10.9` binary of ubuntu 14.04

This would make slightly more sense if the callback was `fb_cb_icon`, which has a `PurpleBuddy` in `FbDataImagePrivate->data`, and it sorta makes sense that a xmlnode_free of the blist save callback would free it. Sorta.

The above is probably nonsense.

I really wish we could reproduce this bug.

**Comment by rrthomas** [comment]
Here's the full log, with pidgin-dbg installed:

    ==5828== Invalid read of size 8
    ==5828==    at 0xEF1792C: fb_api_message_free (api.c:3004)
    ==5828==    by 0x760E6B7: g_slist_foreach (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x760E6DA: g_slist_free_full (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0xEF1A7EC: fb_cb_image (facebook.c:456)
    ==5828==    by 0xEF19783: fb_data_image_cb (data.c:547)
    ==5828==    by 0xEF27ED1: purple_http_connection_terminate (http.c:1688)
    ==5828==    by 0xEF26CED: _purple_http_recv_loopbody (http.c:1260)
    ==5828==    by 0xEF26D46: _purple_http_recv (http.c:1271)
    ==5828==    by 0x17C4CD: pidgin_io_invoke (gtkeventloop.c:73)
    ==5828==    by 0x75F1CE4: g_main_context_dispatch (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x75F2047: g_main_context_iterate.isra.24 (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x75F2309: g_main_loop_run (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x61F8446: gtk_main (in /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0.2400.23)
    ==5828==    by 0x143368: main (gtkmain.c:880)
    ==5828==  Address 0x1d3e7470 is 32 bytes inside a block of size 88 free'd
    ==5828==    at 0x4C2BDEC: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
    ==5828==    by 0x795AA4B: xmlnode_free (xmlnode.c:327)
    ==5828==    by 0x795AA4B: xmlnode_free (xmlnode.c:327)
    ==5828==    by 0x795AA4B: xmlnode_free (xmlnode.c:327)
    ==5828==    by 0x795AA4B: xmlnode_free (xmlnode.c:327)
    ==5828==    by 0x795AA4B: xmlnode_free (xmlnode.c:327)
    ==5828==    by 0x78FCFC0: purple_blist_sync (blist.c:386)
    ==5828==    by 0x78FD218: save_cb (blist.c:392)
    ==5828==    by 0x75F2702: g_timeout_dispatch (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x75F1CE4: g_main_context_dispatch (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x75F2047: g_main_context_iterate.isra.24 (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x75F2309: g_main_loop_run (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x61F8446: gtk_main (in /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0.2400.23)
    ==5828==    by 0x143368: main (gtkmain.c:880)
    ==5828== 
    ==5828== Invalid free() / delete / delete[] / realloc()
    ==5828==    at 0x4C2BDEC: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
    ==5828==    by 0xEF17943: fb_api_message_free (api.c:3005)
    ==5828==    by 0x760E6B7: g_slist_foreach (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x760E6DA: g_slist_free_full (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0xEF1A7EC: fb_cb_image (facebook.c:456)
    ==5828==    by 0xEF19783: fb_data_image_cb (data.c:547)
    ==5828==    by 0xEF27ED1: purple_http_connection_terminate (http.c:1688)
    ==5828==    by 0xEF26CED: _purple_http_recv_loopbody (http.c:1260)
    ==5828==    by 0xEF26D46: _purple_http_recv (http.c:1271)
    ==5828==    by 0x17C4CD: pidgin_io_invoke (gtkeventloop.c:73)
    ==5828==    by 0x75F1CE4: g_main_context_dispatch (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x75F2047: g_main_context_iterate.isra.24 (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x75F2309: g_main_loop_run (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x61F8446: gtk_main (in /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0.2400.23)
    ==5828==    by 0x143368: main (gtkmain.c:880)
    ==5828==  Address 0x1d3e7450 is 0 bytes inside a block of size 88 free'd
    ==5828==    at 0x4C2BDEC: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
    ==5828==    by 0x795AA4B: xmlnode_free (xmlnode.c:327)
    ==5828==    by 0x795AA4B: xmlnode_free (xmlnode.c:327)
    ==5828==    by 0x795AA4B: xmlnode_free (xmlnode.c:327)
    ==5828==    by 0x795AA4B: xmlnode_free (xmlnode.c:327)
    ==5828==    by 0x795AA4B: xmlnode_free (xmlnode.c:327)
    ==5828==    by 0x78FCFC0: purple_blist_sync (blist.c:386)
    ==5828==    by 0x78FD218: save_cb (blist.c:392)
    ==5828==    by 0x75F2702: g_timeout_dispatch (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x75F1CE4: g_main_context_dispatch (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x75F2047: g_main_context_iterate.isra.24 (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x75F2309: g_main_loop_run (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x61F8446: gtk_main (in /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0.2400.23)
    ==5828==    by 0x143368: main (gtkmain.c:880)

**Comment by dequis** [comment]
So the good news is that my guess was right

The bad news is that my guess was right.

What the hell.

Are there no valgrind messages before these?

**Comment by rrthomas** [comment]
There's a bunch of "Conditional jump or move depends on uninitialised values" from the guts of libuim-scm, and Cairo/libpixman, which I'm guessing is irrelevant. Then:

    ==5828== Invalid read of size 8
    ==5828==    at 0xEF1A76B: fb_cb_image (facebook.c:448)
    ==5828==    by 0xEF19783: fb_data_image_cb (data.c:547)
    ==5828==    by 0xEF27ED1: purple_http_connection_terminate (http.c:1688)
    ==5828==    by 0xEF26CED: _purple_http_recv_loopbody (http.c:1260)
    ==5828==    by 0xEF26D46: _purple_http_recv (http.c:1271)
    ==5828==    by 0x17C4CD: pidgin_io_invoke (gtkeventloop.c:73)
    ==5828==    by 0x75F1CE4: g_main_context_dispatch (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x75F2047: g_main_context_iterate.isra.24 (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x75F2309: g_main_loop_run (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x61F8446: gtk_main (in /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0.2400.23)
    ==5828==    by 0x143368: main (gtkmain.c:880)
    ==5828==  Address 0x1d3e7470 is 32 bytes inside a block of size 88 free'd
    ==5828==    at 0x4C2BDEC: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
    ==5828==    by 0x795AA4B: xmlnode_free (xmlnode.c:327)
    ==5828==    by 0x795AA4B: xmlnode_free (xmlnode.c:327)
    ==5828==    by 0x795AA4B: xmlnode_free (xmlnode.c:327)
    ==5828==    by 0x795AA4B: xmlnode_free (xmlnode.c:327)
    ==5828==    by 0x795AA4B: xmlnode_free (xmlnode.c:327)
    ==5828==    by 0x78FCFC0: purple_blist_sync (blist.c:386)
    ==5828==    by 0x78FD218: save_cb (blist.c:392)
    ==5828==    by 0x75F2702: g_timeout_dispatch (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x75F1CE4: g_main_context_dispatch (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x75F2047: g_main_context_iterate.isra.24 (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x75F2309: g_main_loop_run (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x61F8446: gtk_main (in /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0.2400.23)
    ==5828==    by 0x143368: main (gtkmain.c:880)
    ==5828== 
    ==5828== Invalid write of size 8
    ==5828==    at 0xEF1A794: fb_cb_image (facebook.c:449)
    ==5828==    by 0xEF19783: fb_data_image_cb (data.c:547)
    ==5828==    by 0xEF27ED1: purple_http_connection_terminate (http.c:1688)
    ==5828==    by 0xEF26CED: _purple_http_recv_loopbody (http.c:1260)
    ==5828==    by 0xEF26D46: _purple_http_recv (http.c:1271)
    ==5828==    by 0x17C4CD: pidgin_io_invoke (gtkeventloop.c:73)
    ==5828==    by 0x75F1CE4: g_main_context_dispatch (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x75F2047: g_main_context_iterate.isra.24 (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x75F2309: g_main_loop_run (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x61F8446: gtk_main (in /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0.2400.23)
    ==5828==    by 0x143368: main (gtkmain.c:880)
    ==5828==  Address 0x1d3e7470 is 32 bytes inside a block of size 88 free'd
    ==5828==    at 0x4C2BDEC: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
    ==5828==    by 0x795AA4B: xmlnode_free (xmlnode.c:327)
    ==5828==    by 0x795AA4B: xmlnode_free (xmlnode.c:327)
    ==5828==    by 0x795AA4B: xmlnode_free (xmlnode.c:327)
    ==5828==    by 0x795AA4B: xmlnode_free (xmlnode.c:327)
    ==5828==    by 0x795AA4B: xmlnode_free (xmlnode.c:327)
    ==5828==    by 0x78FCFC0: purple_blist_sync (blist.c:386)
    ==5828==    by 0x78FD218: save_cb (blist.c:392)
    ==5828==    by 0x75F2702: g_timeout_dispatch (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x75F1CE4: g_main_context_dispatch (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x75F2047: g_main_context_iterate.isra.24 (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x75F2309: g_main_loop_run (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0)
    ==5828==    by 0x61F8446: gtk_main (in /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0.2400.23)
    ==5828==    by 0x143368: main (gtkmain.c:880)

**Comment by dequis** [comment]
Do you have any other third party plugins installed / enabled?

Do you have a moderately recent intel cpu?

**Comment by rrthomas** [comment]
The only 3rd-party plugin I can find (going by version number different from Pidgin's) is Ubuntu Indicator (provides an indicator icon). My CPU is Sandy Bridge.

**Comment by dequis** [comment]
>My CPU is Sandy Bridge.

Aw yeah.

Get the x86_64 deb from here:

https://github.com/mozilla/rr/releases

Run this to test that everything is okay:

```
rr record ls
rr replay
```

And type "c" in gdb. If that replays the ls command successfully, try this:

```
PURPLE_UNSAFE_DEBUG=1 PURPLE_VERBOSE_DEBUG=1 rr record pidgin --debug
```

And make it crash. I recommend disabling all other accounts before trying this.

Then ensure everything is fine by doing another `rr replay` followed by `c`. If that gets to the end without errors (rather, without more errors than the one you caught) then we can have a better look at the issue.

**Comment by rrthomas** [comment]
The test with `ls` worked fine, so started pidgin. It seems to be using a lot of CPU, but it's hard to tell anything more as whatever it's doing has made `top` fail to start up. So I guess I'll just leave it running and see what happens.

**Comment by rrthomas** [comment]
I noticed in the initial test a recommendation that I change my CPU governor to `performance` (it was on `powersave`), and so I did. Last Pidgin log message is from just after I started up, about 15 minutes ago.

**Comment by dequis** [comment]
Well, uh, it's supposed to be nearly as fast as running it without rr, and definitely much faster than valgrind. Hanging for 15 minutes doesn't seem okay. The CPU governor never seemed to be a big deal for me but i'm on haswell. Maybe you're low on ram?

**Comment by rrthomas** [comment]
Hmm "rr works on real applications such as Firefox, with low overhead". And Pidgin still hasn't started up after 20 minutes.

**Comment by rrthomas** [comment]
Nope, I have 9Gb "available" out of 16Gb total. And the system, despite being under a load of over 4, is perfectly responsive.

**Comment by dequis** [comment]
Well, ctrl-c that stuff, maybe try again a couple of times.

**Comment by rrthomas** [comment]
OK, it hangs every time after "prefs: /pidgin/conversations/placement changed, scheduling save."

I can't Ctrl+C, I have to `kill -9` the rr process.

**Comment by rrthomas** [comment]
There seems to be a problem with rr on my system: if I try starting `rr record emacs -Q`, nothing comes up, though it doesn't stop `top` working as with `pidgin`. If I try `rr record gthumb` then gthumb does open, but I get a message saying some system call didn't work as expected, and I should start gthumb under eog and then attach rr to that.

I'm using a stock Ubuntu 3.19 kernel, but maybe there's something incompatible with the latest rr there…

**Comment by dequis** [comment]
Welp, there's not much i can do with that. Using rr is YMMV, there are plenty of subtle bugs everywhere, especially with less tested microarchitectures or programs. If you're feeling bored submit a bug to https://github.com/mozilla/rr/issues/new and maybe you'll help them out with that.

You could probably do `gdb -p $(pidof rr)` when it's hung and then "bt" to see what it's doing, but i'm not even sure if that's the correct way to debug rr.

I don't want to go further down this rabbit hole, so i'll try to think of something else.

**Comment by rrthomas** [comment]
rr sounds really cool, but I don't have the time to spare to help with it right now; my current bug-hunting time is going on purple-facebook! Do let me know anything else I can do.

**Comment by dequis** [comment]
Okay let's try filling the thing with breakpoints and prints

Put this in a file called `input.txt`

```
set logging on
set pagination off
set print repeats unlimited
set print elements 0x1000
set breakpoint pending on
b facebook/data.c:543
commands
bt
p img
p priv
p *priv
p driv
p *driv
p driv->msgs->length
p g_hash_table_size(driv->imgs)
c
end
b facebook/data.c:448
commands
bt
p func
p img
p priv
p *priv
p fata->priv
p *fata->priv
p fata->priv->msgs->length
p g_hash_table_size(fata->priv->imgs)
c
end
b fb_data_image_queue
commands
bt
p fata->priv
p *fata->priv
p fata->priv->msgs->length
p g_hash_table_size(fata->priv->imgs)
c
end
run
```

Run `gdb -x input.txt pidgin`. No extra debug parameters, this will be enough noise.

It will write its output to a gdb.txt file in the current directory. Post that.

**Comment by rrthomas** [comment]
It's 1.4Mb. Where would you like it?

**Comment by dequis** [comment]
Oh my. That's going to take a while to read through. You can zip it and attach it right here in this ticket.

**Comment by rrthomas** [comment]
[gdb.zip](https://github.com/jgeboski/purple-facebook/files/100293/gdb.zip)

**Comment by dequis** [comment]
Still having this issue? I imagine it's the kind of thing that fixes itself once you mark something as read.

Do you have a groupchat that includes someone who isn't in your friends list?

Do you have a groupchat in which the same image was repeated several times?

**Comment by dequis** [comment]
Finally managed to reproduce it. Needed to receive images from someone who isn't in my friends list.

Better valgrind:

```
==29615== Invalid read of size 8
==29615==    at 0x2397FFE4: fb_cb_image (facebook.c:448)
==29615==    by 0x2397EC09: fb_data_image_cb (data.c:547)
==29615==    by 0x23989F2D: purple_http_connection_terminate (http.c:1688)
==29615==    by 0x2398C2BE: _purple_http_recv_loopbody.isra.24 (http.c:1260)
==29615==    by 0x2398C887: _purple_http_recv (http.c:1271)
==29615==    by 0x19198B: pidgin_io_invoke (gtkeventloop.c:73)
==29615==    by 0x7484E8E: g_main_dispatch (gmain.c:3154)
==29615==    by 0x7485DC5: g_main_context_dispatch (gmain.c:3769)
==29615==    by 0x7485FC9: g_main_context_iterate (gmain.c:3840)
==29615==    by 0x7486469: g_main_loop_run (gmain.c:4034)
==29615==    by 0x6043786: gtk_main (in /usr/lib/libgtk-x11-2.0.so.0.2400.29)
==29615==    by 0x1B65EA: main (gtkmain.c:937)
==29615==  Address 0x2a350770 is 32 bytes inside a block of size 40 free'd
==29615==    at 0x4C29D2A: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==29615==    by 0x748DDFD: g_free (gmem.c:189)
==29615==    by 0x74AA95F: g_slist_foreach (gslist.c:878)
==29615==    by 0x74A9FF2: g_slist_free_full (gslist.c:172)
==29615==    by 0x239815EC: fb_cb_api_contact (facebook.c:167)
==29615==    by 0x71E9359: g_cclosure_marshal_VOID__POINTERv (gmarshal.c:2026)
==29615==    by 0x71E3F54: _g_closure_invoke_va (gclosure.c:864)
==29615==    by 0x7201098: g_signal_emit_valist (gsignal.c:3292)
==29615==    by 0x7202492: g_signal_emit_by_name (gsignal.c:3479)
==29615==    by 0x2397D5F9: fb_api_cb_contact (api.c:2006)
==29615==    by 0x23989F2D: purple_http_connection_terminate (http.c:1688)
==29615==    by 0x2398C2BE: _purple_http_recv_loopbody.isra.24 (http.c:1260)
==29615==  Block was alloc'd at
==29615==    at 0x4C28C10: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==29615==    by 0x748DC7B: g_malloc (gmem.c:94)
==29615==    by 0x74AB353: g_memdup (gstrfuncs.c:391)
==29615==    by 0x2397B46A: fb_api_message_dup (api.c:2979)
==29615==    by 0x2397FE0C: fb_cb_api_messages (facebook.c:508)
==29615==    by 0x239815DD: fb_cb_api_contact (facebook.c:166)
==29615==    by 0x71E9359: g_cclosure_marshal_VOID__POINTERv (gmarshal.c:2026)
==29615==    by 0x71E3F54: _g_closure_invoke_va (gclosure.c:864)
==29615==    by 0x7201098: g_signal_emit_valist (gsignal.c:3292)
==29615==    by 0x7202492: g_signal_emit_by_name (gsignal.c:3479)
==29615==    by 0x2397D5F9: fb_api_cb_contact (api.c:2006)
==29615==    by 0x23989F2D: purple_http_connection_terminate (http.c:1688)
```

At least two calls to `fb_cb_api_contact` are involved here, although I got like 12 times. That function is the callback for when the plugin requested additional information about a non-friend contact and receives the result.

Near the end it does this:

```c
163             msgs = fb_data_take_messages(fata, user->uid);
164
165             if (msgs != NULL) {
166                     fb_cb_api_messages(api, msgs, fata);
167                     g_slist_free_full(msgs, (GDestroyNotify) fb_api_message_free);
168             }
```

Which takes some queued messages matching this user id and processes them. The messages are taken (removed and returned) from `fata->priv->msgs`, passed to `fb_cb_api_messages`, and then freed.

Assuming no other contacts sent messages, `fata->priv->msgs` has length 0 after `fb_data_take_messages` returns.

In `fb_cb_api_messages` there's this:

```c
506                     if (msg->flags & FB_API_MESSAGE_FLAG_IMAGE) {
507                             if (!(msg->flags & FB_API_MESSAGE_FLAG_DONE)) {
508                                     msg = fb_api_message_dup(msg, TRUE);
509                                     fb_data_image_add(fata, msg->text,
510                                                       fb_cb_image, msg);
511                                     fb_data_add_message(fata, msg);
512                                     fb_data_image_queue(fata);
513                                     continue;
```

Which:

* `fb_api_message_dup` - duplicates the message so that it will stay around after it's freed
* `fb_data_image_add` - adds it to the queue of images to be requested. References to `msg` and `msg->text` are passed, and this function expects them to stay around until `fb_cb_image` is called (that's where it crashes)
* `fb_data_add_message` - adds the message back to `fata->priv->msgs`

So `fata->priv->msgs` has length 1 now.

Soon later, `fb_cb_api_contact` is called again because someone sent a bit too many http requests asking for extra information about that contact. The above is repeated. The contents of `fata->priv->msgs` are freed and new copies are added back to it. The copies done before were supposed to be kept around.

A few seconds later, the first image download finishes. `fb_cb_image` is called, which tries to access the first copy of `msg` from the first `fb_data_image_add` call, and since it was freed in the second iteration, things break.

**Comment by rrthomas** [comment]
Well done! Sorry I didn't reply to the earlier questions in time, but you were quite right: I'm still having the issue (it doesn't clear, just repeats each time). I have indeed a group chat in which there are people not in my friends. I don't know whether I was getting repeated images, I think it crashed before displaying any.

**Comment by jgeboski** [comment]
This has been resolved as of [92885e0](https://bitbucket.org/pidgin/main/commits/92885e0456edad1a04d6dedb418e5d86bd7a3be6), and will appear in the next release.

**Comment by rrthomas** [comment]
Great stuff, thanks very much!

**Comment by jgeboski** [comment]
This patch is in the latest release, [92885e0456ed](https://github.com/jgeboski/purple-facebook/releases/tag/92885e0456ed). Test away.

**Comment by dequis** [comment]
Also huge thanks to @rrthomas for the help debugging this issue. The hint i needed was the `fb_cb_api_contact` that called `fb_cb_api_messages`, which is something that never appeared before until that last gdb log.

>Sorry I didn't reply to the earlier questions in time

Don't worry, I managed to reproduce the bug like 10 mins after posting my first comment with the questions.

>I don't know whether I was getting repeated images, I think it crashed before displaying any.

That was just a side effect of this bug, the image download was started several times and it crashed once the first one finished. I thought it could be the *cause*, but the other one was.

**Comment by rrthomas** [comment]
Looking fine so far, but the Chat of Doom™ appears to have expired (it's not reopening automatically), so I suppose I can't really tell.

**Comment by gaborigloi** [comment]
Thanks for fixing this, now it works perfectly! :) Pidgin successfully received and showed unread messages from non-friends that contain images, from the group chat mentioned in https://github.com/jgeboski/purple-facebook/issues/165 , without crashing.
